### PR TITLE
Some device list tests

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 1]
```diff
@@ -407,3 +480,296 @@ sub is_user_in_changed_list
...
+         assert_json_keys( $body, qw( changed ) );
+
+         my $changed = $body->{changed};
+
+         any { $remote_user->user_id eq $_ } @{ $changed }
+            or die "user not in changed list";
+
+         Future->done(1);
+      });
+   };
```
Comment (richvdh) [inline tests/41end-to-end-keys/06-device-lists.pl]: trailing NL would be nice

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 2]
```diff
@@ -407,3 +480,296 @@ sub is_user_in_changed_list
...
+            ->then( sub {
+               my ( $body ) = @_;
+
+               log_if_fail "First body", $body;
+
+               Future->done( is_user_in_changed_list( $remote_leaver, $body ) );
+            })
+         };
+      })->then( sub {
+         matrix_sync_again( $creator )
```
Comment (richvdh) [inline tests/41end-to-end-keys/06-device-lists.pl]: if you're just after the next_batch you can do `$creator->sync_next_batch`.

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 3]
```diff
@@ -407,3 +480,296 @@ sub is_user_in_changed_list
...
+      })->then( sub {
+         my ( $body ) = @_;
+
+         $to_token = $body->{to_batch};
```
Comment (richvdh) [inline tests/41end-to-end-keys/06-device-lists.pl]: erm; this is news to me. I thought we were supposed to use the `next_batch` token for `/keys/changes`.

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 4]
```diff
@@ -407,3 +480,296 @@ sub is_user_in_changed_list
...
+            ->then( sub {
+               my ( $body ) = @_;
+
+               log_if_fail "First body", $body;
+
+               Future->done( is_user_in_changed_list( $other_user, $body ) );
+            })
+         };
+      })->then( sub {
+         matrix_sync_again( $creator )
```
Comment (richvdh) [inline tests/41end-to-end-keys/06-device-lists.pl]: again, use `$creator->sync_next_batch`

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 5]
```diff
@@ -407,3 +480,296 @@ sub is_user_in_changed_list
...
```
Comment (richvdh) [inline tests/41end-to-end-keys/06-device-lists.pl]: `$creator->sync_next_batch`

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 6]
```diff
@@ -407,3 +480,296 @@ sub is_user_in_changed_list
...
+            ->then( sub {
+               my ( $body ) = @_;
+
+               log_if_fail "First body", $body;
+
+               Future->done( is_user_in_changed_list( $remote_user, $body ) );
+            })
+         };
+      })->then( sub {
+         matrix_sync_again( $creator )
```
Comment (richvdh) [inline tests/41end-to-end-keys/06-device-lists.pl]: `$creator->sync_next_batch`

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 7]
```diff
@@ -227,6 +238,68 @@ sub is_user_in_changed_list
...
+         }
+      })->then( sub {
+         matrix_join_room( $remote_leaver, $room_id );
+      })->then( sub {
+         repeat_until_true {
+            matrix_sync_again( $creator, timeout => 1000 )
+            ->then( sub {
+               my ( $body ) = @_;
+
+               log_if_fail "Third body", $body;
```
Comment (richvdh) [inline tests/41end-to-end-keys/06-device-lists.pl]: you've been hiding the bodies again.

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 8]
```diff
@@ -407,3 +480,296 @@ sub is_user_in_changed_list
...
+            matrix_sync_again( $creator, timeout => 1000 )
+            ->then( sub {
+               my ( $body ) = @_;
+
+               log_if_fail "Second body", $body;
+
+               Future->done( is_user_in_left_list( $other_user, $body ) );
+            })
+         };
+      });
```
Comment (richvdh) [inline tests/41end-to-end-keys/06-device-lists.pl]: should we have a test for /keys/changes here?

File: tests/41end-to-end-keys/06-device-lists.pl [hunk 9]
```diff
@@ -407,3 +480,296 @@ sub is_user_in_changed_list
...
```
Comment (richvdh) [inline tests/41end-to-end-keys/06-device-lists.pl]: /keys/changes?

**Comment by erikjohnston** [comment]
I've added `/keys/changes` tests to the end of the `/sync` tests, as they would be practically identical.

**Pull request merged**